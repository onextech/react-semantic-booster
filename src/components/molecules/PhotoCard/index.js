import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Header, Button, Image } from 'semantic-ui-react';


const getCardLinkButtonCss = () => `
    &.ui.button, & {
      width: 100%;
      height: 260px;
      display: flex;
      position: relative;
      padding: 30px 4.65116%;
      flex-direction: column;
      overflow: hidden;
      text-align: center;
      text-decoration: none;
      color: rgb(255, 255, 255);
      background-color: rgb(31, 31, 31);
      
      &:hover,
      &:active, 
      &:focus {
        img.animated {
          opacity: 0.7;
          &.scale {
            transform: scale(1.05);
          }
        }
      }
      
      align-items: flex-start;
      justify-content: flex-end;
      text-align: left;
      
      &.centered {
        align-items: center;
        justify-content: center;
        text-align: center;
      }
    }
  `;
const CardLink = styled(Link)`${getCardLinkButtonCss()}`;
const CardButton = styled(Button)`${getCardLinkButtonCss()}`;

const CardContent = styled.div`
    position: relative;
    max-width: 520px;
    z-index: 2;
  `;

const CardImage = styled(Image)`
    &.ui.image {
      width: 100%;
      height: 100%;
      display: block;
      position: absolute;
      z-index: 1;
      left: 0;
      top: 0;
      opacity: 0.4;
      object-fit: cover;
      transition: opacity 800ms ease-in-out, transform 800ms ease-in-out;
    }
  `;

const PhotoCard = ({
  content,
  cardProps,
  imgProps,
  headerProps,
  centered,
  animated,
  onClick,
}) => {
  const getCardContainerClass = () => {
    const cardContainerClass = [];
    if (centered) {
      cardContainerClass.push('centered');
    }
    return cardContainerClass.join(' ');
  };

  const getCardImageClass = () => {
    const cardImageClass = [];
    if (animated) {
      cardImageClass.push('animated');
      cardImageClass.push(animated);
    }
    return cardImageClass.join(' ');
  };

  const handleClick = (e, data) => onClick(e, data);

  const CardContainer = onClick ? CardButton : CardLink;

  return (
    <div style={{ backgroundColor: 'black' }}>
      <CardContainer
        className={getCardContainerClass()}
        onClick={handleClick}
        to="#"
        {...cardProps}>
        {
          content &&
          <CardContent>
            <Header as="h4" {...headerProps}>
              {
                content.subheader &&
                <Header.Subheader style={{ opacity: '0.8' }}>
                  {content.subheader}
                </Header.Subheader>
              }
              {content.header}
            </Header>
          </CardContent>
        }
        <CardImage className={getCardImageClass()} {...imgProps} />
      </CardContainer>
    </div>
  );
};

PhotoCard.propTypes = {
  content: PropTypes.shape({
    header: PropTypes.string,
    subheader: PropTypes.string,
  }),
  cardProps: PropTypes.object, // props to lay on the card
  imgProps: PropTypes.object, // <img> props
  headerProps: PropTypes.object, // semantic-ui header props
  centered: PropTypes.bool,
  animated: PropTypes.string, // enums for css animations: 'scale'
  onClick: PropTypes.func,
};

PhotoCard.defaultProps = {
  content: undefined,
  centered: undefined,
  animated: undefined,
  onClick: undefined,
  cardProps: undefined,
  imgProps: undefined,
  headerProps: undefined,
};

export default PhotoCard;
