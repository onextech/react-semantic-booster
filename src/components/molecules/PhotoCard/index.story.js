import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Grid from '../../atoms/Grid';
import PhotoCard from '.';
import { testImage1 } from '../../../../test/placeholders';


export const photoCardExamples = [
  {
    content: {
      header: 'hello world',
      subheader: 'foo',
    },
    key: 'hello-world',
    imgProps: {
      src: testImage1,
    },
    to: '/link-1',
  },
  {
    content: {
      header: 'hello world',
      subheader: 'bar',
    },
    key: 'hello-world-2',
    imgProps: {
      src: testImage1,
    },
    to: '/link-2',
  },
  {
    content: {
      header: 'hello world',
      subheader: 'baz',
    },
    key: 'hello-world-3',
    imgProps: {
      src: testImage1,
    },
    to: '/link-3',
  },
];

export const firstPhotoCard = photoCardExamples[0];

storiesOf('PhotoCard', module)
  .add('Default', () => (
    <PhotoCard
      onClick={action('clicked')}
      animated="scale"
      content={firstPhotoCard.content}
      cardProps={{ card: firstPhotoCard.key }}
      headerProps={{ inverted: true, size: 'large' }}
      imgProps={firstPhotoCard.imgProps} />
  ))
  .add('Link', () => (
    <PhotoCard
      centered
      animated="scale"
      content={firstPhotoCard.content}
      cardProps={{ card: firstPhotoCard.key, to: firstPhotoCard.to }}
      headerProps={{ inverted: true, size: 'large' }}
      imgProps={firstPhotoCard.imgProps} />
  ))
  .add('Gridded', () => (
    <div>
      <Grid container columns="equal" stackable>
        <Grid.Row>
          {photoCardExamples.map(photoCardExample => (
            <Grid.Column key={photoCardExample.key}>
              <PhotoCard
                animated="scale"
                content={photoCardExample.content}
                cardProps={{ card: photoCardExample.key, to: photoCardExample.to }}
                headerProps={{ inverted: true, size: 'large' }}
                imgProps={photoCardExample.imgProps} />
            </Grid.Column>
          ))}
        </Grid.Row>
      </Grid>
      <Grid container columns="equal" fluid stackable attached>
        <Grid.Row>
          {photoCardExamples.slice(0, 2).map(photoCardExample => (
            <Grid.Column key={photoCardExample.key}>
              <PhotoCard
                animated="scale"
                content={photoCardExample.content}
                cardProps={{ card: photoCardExample.key, to: photoCardExample.to }}
                headerProps={{ inverted: true, size: 'large' }}
                imgProps={photoCardExample.imgProps} />
            </Grid.Column>
          ))}
        </Grid.Row>
      </Grid>
    </div>
  ));
