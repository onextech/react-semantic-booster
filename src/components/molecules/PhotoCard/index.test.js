import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';
import { action } from '@storybook/addon-actions';
import PhotoCard from '.';
import { firstPhotoCard } from './index.story';


describe('<PhotoCard>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Router><PhotoCard
        onClick={action('clicked')}
        animated="scale"
        content={firstPhotoCard.content}
        cardProps={{ card: firstPhotoCard.key, fluid: true }}
        headerProps={{ inverted: true, size: 'large' }}
        imgProps={firstPhotoCard.imgProps} />
      </Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
