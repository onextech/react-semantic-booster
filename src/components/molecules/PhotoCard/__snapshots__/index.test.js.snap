// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoCard> renders correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
    }
  }
>
  <button
    card="hello-world"
    className="ui fluid button sc-bdVaJa dvglne"
    disabled={undefined}
    onClick={[Function]}
    role="button"
    tabIndex={undefined}
    to="#"
  >
    <div
      className="sc-bwzfXH crJjvc"
    >
      <h4
        className="ui large inverted header"
      >
        <div
          className="sub header"
          style={
            Object {
              "opacity": "0.8",
            }
          }
        >
          foo
        </div>
        hello world
      </h4>
    </div>
    <img
      className="ui image animated scale sc-htpNat btySty"
      src="https://upload.wikimedia.org/wikipedia/commons/c/cd/Panda_Cub_from_Wolong,_Sichuan,_China.JPG"
    />
  </button>
</div>
`;
